### ISSUE TITLE
[Feature] reconsider prepare lifecycle script?

### ISSUE THREAD

**Issue by typicode** [yarnpkg/berry#2967]
- [ ] I'd be willing to implement this feature ([contributing guide](https://yarnpkg.com/advanced/contributing))
- [ ] This feature is important to have in this repository; a contrib plugin wouldn't do

**Describe the user story**

As a developer, I may need to run some development only postinstall script.

**Describe the solution you'd like**

Enable `prepare` script.

**Describe the drawbacks of your solution**

I'm aware Yarn 2 tries to simplify and limit the number of lifecycle hooks. I think it's good a thing.

However, except with some tricks and AFAIK there's no way to have a development only `postinstall` script with Yarn 2.

**Describe alternatives you've considered**

Since only `postinstall` can be used to run commands after `yarn`, I've created [pinst](https://github.com/typicode/pinst) as a way to disable it when a package is published (simulating a dev-only postinstall script).

There's an example here:
https://typicode.github.io/husky/#/?id=yarn-2

**Comment by sachinraja** [comment]
I believe I've solved this with a simple plugin, [sachinraja/yarn-plugin-postinstall-dev](https://github.com/sachinraja/yarn-plugin-postinstall-dev). This should run only when the developers run `yarn install`.

**Comment by merceyz** [comment]
With https://github.com/yarnpkg/berry/pull/3146 merged `pinst` should be able to remove the `postinstall` script before packing again, we could potentially support overriding lifecycle scripts in [`publishConfig`](https://yarnpkg.com/configuration/manifest#publishConfig) to have a development only `postinstall` as well

**Comment by rxliuli** [comment]
> I believe I've solved this with a simple plugin, [yarn-plugin-postinstall-dev](https://github.com/sachinraja/yarn-plugin-postinstall-dev). This should run only when the developers run `yarn install`. Though this should definitely be supported without plugins.

There are some errors in this script, but I modified it slightly and it can be used normally.

**Comment by sachinraja** [comment]
@rxliuli Please create an issue on the repo as this is not the right place to report issues with it

**Comment by paescuj** [comment]
I agree with @typicode that Yarn's initiative to simplify and limit the number of lifecycle hooks is generally a good thing.
But IMO the "prepare" lifecycle script is pretty essential.
The docs of Yarn itself are recommending to avoid the "postinstall" script whenever possible: https://yarnpkg.com/advanced/lifecycle-scripts/#a-note-about-postinstall
However, without the "prepare" script there is currently no (clean) alternative for tools like [husky](https://github.com/typicode/husky) or [simple-git-hooks](https://github.com/toplenboren/simple-git-hooks).

I'd really appreciate some thoughts / feedback from the maintainers here! I'll be happy to come up with a PR if this request is accepted!

**Comment by WoodyWoodsta** [comment]
To add a use case to this issue: we share git hooks among the team, which perform some important checks before commiting. In order to make these checks more watertight, it's best that it's not left up to the development to configure the hooks path. With NPM, we would use the `prepare` lifecycle hook as the place to setup the git hook changes, but we now no longer have an out-the-box solution.

**Comment by macbem** [comment]
I second this. It would be useful to bootstrap auth for our private package registries since the `getNpmAuthenticationHeader` plugin lifecycle hook is ran in parallel and it's not feasible to add sign-in logic there. It would be useful to have a `prepare` hook that would allow us to run sign-in logic and tokens would be available for extraction when packages are requested. Without this lifecycle script, the only viable option I see is to wrap `yarn install` with a custom script which doesn't sound like a good solution tbh.

**Comment by Devidian** [comment]
I like yarn, i like [typia](https://github.com/samchon/typia/issues/1106), i like to work with latest updates. Sadly typia needs prepare script and berry doesn't support it. So typia doesn't support berry. That makes me sad, because i cant stick up to latest version.



Oh well, that is maybe just a +1 with an usage example from my side.

**Comment by blimmer** [comment]
This is a bigger problem now that NPM has broken the `pinst` workflow, which is what everyone using `husky` on Yarn Berry uses.

See https://github.com/typicode/pinst/issues/23 / https://gist.github.com/djcsdy/2f5a287b3ba16f2a8f0312f45588e6ce

TLDR; we really need a lifecycle script that only runs when running `yarn install` on the package in question.

**Comment by dhouck** [comment]
NPM is finally no longer running postinstall scripts by default, but both PNPM and NPM loudly warn about them, so developer-only scripts *still* canʼt safely go in postinstall.  And with pinst being… inconsistent… I still donʼt think thereʼs an actual solution to this.
